### ISSUE TITLE
New Feature: 4D Tensor Input for LSTM RNN Neural Network

### ISSUE THREAD

**Issue by arita297** [tensorflow/tensorflow#2560]
Currently TF RNN LSTM accepts this tensor rank:   (ht, mt)  of rank 1

     (ht_o, mt_o) = LSTM(ht, mt, W)
     with   ht size is (1 to d)   ,  same for mt

Would it be possible to accept rank 2 or rank 3 for the input state of RNN LSTM
    ht size  (1 to m, 1 to s, 1 to d)      : Rank 3 

It would allows Multi-dimensionnal input such as :
    Input  size  (NSample, Mtype, Nchar, Nsequence) 

Especially, it makes sense for Grid LSTM.

**Comment by Nessphoro** [comment]
RNNs are being reworked,
you're better off implementing them yourself in tensorflow (just follow one of the many github tutorials),
automatic differentiation will take care of computing the correct gradients.

**Comment by ebrevdo** [comment]
@Nessphoro that answer is incorrect.  RNNs are a stable API.

We will add support for higher rank inputs and nested tuples of inputs and outputs in a backwards compatible way in the future (i can't give estimates, but imagine it'll happen this summer sometime)

**Comment by Nessphoro** [comment]
@ebrevdo I've though that some work was being done to change how RNNs and LSTMs work because of https://github.com/tensorflow/tensorflow/blob/82ff4cd8b0d541ede107d34d8eecc769c91dda11/tensorflow/models/rnn/rnn_cell.py
And the stale documentation in master with regards to that.

**Comment by ebrevdo** [comment]
We just moved the code into tf.nn.rnn_cell.*, tf.nn.rnn, tf.nn.dynamic_rnn,
tf.nn.seq2seq, etc.

**Comment by Nessphoro** [comment]
Thanks!
 
The docs on the tutorial for seq2seq learning still refer to the models/rnn namespace still.

**Comment by girving** [comment]
@ebrevdo: Feel free to reassign or mark this contributions welcome if that's best.

**Comment by ebrevdo** [comment]
This is being worked on internally.  I'll see if we have an external github account for the contributor.

**Comment by girving** [comment]
@kosklain Here you go!

**Comment by kosklain** [comment]
@arita297 Hey, the changes are now in (more concretely, [this commit](https://github.com/tensorflow/tensorflow/commit/b5ad739fe9939e953416636ae8e188a5feacbcdf)). As you can see in the docs, you can do what you want by specifying objects that inherit from RNNCell (just like your Grid LSTM would), and then the `output_size` and `state_size` properties can specify multi-dimensional sizes by using tensorflow TensorShapes.

So if you have
`(output, next_state) = YourCell(input, prev_state)`
such that output is size (o1,o2,o3), next_state is (h, c) with both having size (s1, s2, s3), then your properties should look like:

```
 @property
 def state_size(self):
   return (tf.TensorShape((s1, s2, s3)), tf.TensorShape((s1, s2, s3))

@property
def output_size(self):
  return tf.TensorShape(o1, o2, o3)
```

With that, you should be able to operate with multidimensional input/output/states.

**Comment by girving** [comment]
@kosklain Should we close the issue, or is there more to do?

**Comment by kosklain** [comment]
@girving this can now be closed, with that commit the feature should be currently fully supported by all rnn modules that exist in `ops/rnn.py`.

**Comment by arita297** [comment]
@kosklain : Sure, sounds good.
Shall we update from the master branch ?
   -it b.gcr.io/tensorflow/tensorflow

**Comment by kosklain** [comment]
@arita297 I believe docker will only get you at most the 0.8 release. To get the bleeding-edge master branch you should download the sources and compile.

**Comment by arita37** [comment]
@kosklain   : Ok, Sure

Does 0.9c release now contains this ?

https://www.tensorflow.org/versions/r0.9/get_started/os_setup.html

**Comment by arita37** [comment]
Hello,

@kosklain : Ok, Sure

Does 0.10c release now contains this ?

https://www.tensorflow.org/versions/r0.10

Thanks
